### PR TITLE
feat(pipeline): limit requests per second

### PR DESCRIPTION
**Pull request by warren830**
added a config to limit gira, gitlab, feishu collecting rate

# Summary
Added rate limit control in gira, gitlab, feishu plugins

### Description
Limit rate of scraping metrics from a large project to avoid causing increased load

### Does this close any open issues?
Closes issue #931 

### Other Information
Will create a new issue to do some work on UI

### CODE REVIEW

File: plugins/core/plugin_utils.go [hunk 1]
```diff
@@ -41,3 +41,14 @@ const SourceIdError = "Missing or Invalid sourceId"
 const InvalidConnectionError = "Your connection configuration is invalid."
 const UnsetConnectionError = "Your connection configuration is not set."
 const UnmarshallingError = "There was a problem unmarshalling the response"
+
+func GetRateLimitPerSecond(options map[string]interface{}, defaultValue int) (int, error) {
+	rateLimitPerSecond, ok := options["rateLimitPerSecond"].(float64)
```
Comment (klesh) [inline plugins/core/plugin_utils.go:46]: if parameter is missing, return default value
if parameter existed, check if valid
  1. invalid : return error if invalid so user will know he passed in a wrong parameter
  2. valid: return int

File: plugins/github/github.go [hunk 2]
```diff
@@ -58,15 +58,12 @@ func (plugin Github) Execute(options map[string]interface{}, progress chan<- flo
...
 	_ = githubApiClient.SetProxy(config.V.GetString("GITHUB_PROXY"))
-	// GitHub API has very low rate limits, so we cycle through multiple tokens to increase rate limits.
-	// Then we set the ants max worker per second according to the number of tokens we have to maximize speed.
-	tokenCount := len(tokens)
-	// We need this rate limit set to 1 by default since there are only 5000 requests per hour allowed for the github api
-	maxWorkerPerSecond := 1
-	if tokenCount > 0 {
-		maxWorkerPerSecond = tokenCount
+
+	rateLimitPerSecondInt, err := core.GetRateLimitPerSecond(options, 1)
```
[the remark this answers is not in the document]
Comment (warren830) [reply plugins/github/github.go:62]: Ok, here is the modification:
```go
        tokenCount := len(tokens)

	if tokenCount == 0 {
		return fmt.Errorf("owner is required for GitHub execution")
	}

	rateLimitPerSecondInt, err := core.GetRateLimitPerSecond(options, tokenCount)
```  
is this OK?

File: plugins/jenkins/jenkins.go [hunk 3]
```diff
@@ -52,13 +53,28 @@ func (j Jenkins) Execute(options map[string]interface{}, progress chan<- float32
...
 		return fmt.Errorf("Failed to decode options: %v", err)
 	}
+
+	var rateLimitPerSecondInt int
+	rateLimitPerSecondInt, err = core.GetRateLimitPerSecond(options, 15)
+	if err != nil {
+		return err
+	}
+
+	// to keep the progress work properly, we set workerNum = 1, so it will work one by one
```
[the remark this answers is not in the document]
Comment (warren830) [reply plugins/jenkins/jenkins.go:68]: Jenkins has its own progress calculation, if we set more workers, the progress will be messed up
```go
var ctx = context.Background()
	// get all jobs
	var jobs, err = worker.jenkins.GetAllJobs(ctx)
	if err != nil {
		return fmt.Errorf("Failed to get jobs from jenkins: %v", err)
	}
	for index, job := range jobs {
		err = scheduler.Submit(func() error {
			logger.Debug("(worker *JenkinsWorker) Submit", job)
			workerErr := worker.syncJob(ctx, job)
			if workerErr != nil {
				return workerErr
			}
			progress <- float32((index + 1)) / float32(len(jobs))    <------- more wokers cause index to be the same
			return nil
		})
		if err != nil {
			return err
		}
	}
```

[the remark this answers is not in the document]
Comment (klesh) [reply plugins/jenkins/jenkins.go:68]: ok, if that was the case, our limitation wont be working, right?
can we pass the rate limit into jenkins worker or sth? if not ,maybe we skip jenkins for now.

File: plugins/core/plugin_utils.go (hunk 1 above)
Comment (klesh) [reply plugins/core/plugin_utils.go:46]: the current code would alway return an error if ratelimit omitted. this is not what we want.

Comment (warren830) [reply plugins/core/plugin_utils.go]: OK, I modified the issues and resubmitted a new pr

**Review by klesh — APPROVED** [review]

**Pull request merged**